### PR TITLE
[UIMA-5753] Get UIMA compile on Java 9 and higher

### PR DESCRIPTION
**Pull request by reckart**
Several changes allowing a successful build of UIMA using a Java 9+ JDK.

### CODE REVIEW

File: uimaj-core/pom.xml [hunk 1]
```diff
@@ -260,4 +260,36 @@
       </plugin>
     </plugins>
   </build>
+  <profiles>
+    <!-- ************ Java 9 enablement ************** -->
```
Comment (mischor) [inline uimaj-core/pom.xml]: I wonder if we could name this java 9 or 10 enablement?

File: uimaj-core/pom.xml [hunk 2]
```diff
@@ -260,4 +260,36 @@
...
+        <!-- 
+          - To avoid having to add the java.xml.bind module in Java 9 and higher, we simply
+          - directly add the correpsonding dependencies. This seems to be the way to go
+          - anyway since java.xml.bind seems to be bound to be removed from future JDKs.
+          - See: https://stackoverflow.com/questions/43574426/how-to-resolve-java-lang-noclassdeffounderror-javax-xml-bind-jaxbexception-in-j/43574427
+          -->
+        <dependency>
+          <groupId>javax.xml.bind</groupId>
+          <artifactId>jaxb-api</artifactId>
+          <version>2.3.0</version>
```
Comment (mischor) [inline uimaj-core/pom.xml]: Do you think that depending on some earlier version might get rid of that problem with extra new lines creeping in?

Comment (reckart) [reply uimaj-core/pom.xml]: I have no idea if the newline issue is related to JAXB. I rather expect that changes in the XSLT code introduce the newlines because turning off the pretty-print option in one test case fixed it... but well, it might be worth a try.

File: uimaj-core/pom.xml (hunk 1 above)
Comment (reckart) [reply uimaj-core/pom.xml]: `Java 9+` ? ; )

File: uimaj-core/src/main/java/org/apache/uima/analysis_engine/annotator/package.html [hunk 3]
```diff
@@ -22,7 +22,7 @@
 <body>
 <p>The Annotator Interfaces, along with supporting interfaces
 and exception classes.</p>
-<p>The annotator interfaces are as follows:
+<p>The annotator interfaces are as follows:</p>
```
Comment (mischor) [inline uimaj-core/src/main/java/org/apache/uima/analysis_engine/annotator/package.html]: Some sources (e.g. http://www.oracle.com/technetwork/articles/java/index-137868.html ) say to only use the <p> tag, not the </p>.  Also, I vaguely recall one of the javadoc lint checker tools complaining if you used some of these closing tags.  also this https://stackoverflow.com/questions/5260368/which-tag-should-be-used-as-paragraph-separator-in-javadoc says the doclet doesn't like the closing </p>

Comment (reckart) [reply uimaj-core/src/main/java/org/apache/uima/analysis_engine/annotator/package.html]: To embed HTML in a comment, enclose it in single quotes (for inline HTML) or three single quotes to begin a verbatim block (and later again to end it.

File: uimaj-core/src/main/java/org/apache/uima/cas/impl/package.html [hunk 4]
```diff
@@ -30,7 +30,7 @@
   <body>
 <p>Implementation and Low-Level API for the CAS Interfaces.</p>
 <p>Internal APIs. Use these APIs at your own risk. APIs in this package are subject to change without notice, even in minor releases. Use of this package is not supported. If you think you have found a bug in this package, please try to reproduce it with the officially supported APIs before reporting it.</p>
-<hr />
+<hr>
```
Comment (mischor) [inline uimaj-core/src/main/java/org/apache/uima/cas/impl/package.html]: please run the javadoc with java 9 / 10 to see if any of these closing tags get complaints...

Comment (reckart) [reply uimaj-core/src/main/java/org/apache/uima/cas/impl/package.html]: JavaDoc 9/10 (and actually I think even 8) do not like self-closing HTML elements.

File: uimaj-parent/pom.xml [hunk 5]
```diff
@@ -472,6 +455,28 @@
                 </goals>
               </execution>
             </executions>
+            <dependencies>
+              <dependency>
```
Comment (mischor) [inline uimaj-parent/pom.xml]: any way to factor this common set of dependencies into one spot?

Comment (reckart) [reply uimaj-parent/pom.xml]: As UIMA dependencies, we can add them to the parent POM such that all modules inherit them.

However, if individual Maven plugins require these dependencies (such as here), I know of no way to do a single declaration that would affect all plugins. Mind that that each plugin has its own classloader.

**Comment by reckart** [comment]
Already answered them ;)

File: uimaj-core/src/main/java/org/apache/uima/analysis_engine/annotator/package.html (hunk 3 above)
Comment (mischor) [reply uimaj-core/src/main/java/org/apache/uima/analysis_engine/annotator/package.html]: ok, my comment should say that the javadoc tool (at least for java 8, maybe for 9/10) complained if you put in the closing paragraph tag '</p>' .  So these probably should not be added?

Comment (reckart) [reply uimaj-core/src/main/java/org/apache/uima/analysis_engine/annotator/package.html]: Hm... I guess "single quote" was not right and I should have said "backtick" (`)

The closing paragraph tag is not a problem. What JavaDoc 8+ is self-closing elements (`<br/>, <hr/>, etc.`).

File: uimaj-core/src/main/java/org/apache/uima/cas/impl/package.html (hunk 4 above)
Comment (mischor) [reply uimaj-core/src/main/java/org/apache/uima/cas/impl/package.html]: +1

File: uimaj-parent/pom.xml (hunk 5 above)
Comment (mischor) [reply uimaj-parent/pom.xml]: ok

**Comment by reckart** [comment]
I am unsure if it is a wise idea to have the JAXB dependencies in a profile which activates on a Java 9+ JDK or if it would be better to simply have them as general runtime dependencies, even under Java 8. Do you have any opinion on that?

**Comment by reckart** [comment]
**Note:** in case anybody is watching this... the UIMA project presently does not make use of GitHub (and the GitHub/ASF integration). We are basically toying around here to explore what the use of GitHub might contribute to the UIMA project development workflow.

**Comment by mischor** [comment]
re: JAXB dependencies for Java 8 - In the interest of stability, I would leave these for Java 9/10, and not put them in for Java 8.

**Comment by reckart** [comment]
I am a bit worried that it makes the build less reproducible. E.g. distributions or fat JARs would end up having different sets of dependencies depending on which JDK was used to build them.

File: uimaj-core/src/main/java/org/apache/uima/analysis_engine/annotator/package.html (hunk 3 above)
Comment (reckart) [reply uimaj-core/src/main/java/org/apache/uima/analysis_engine/annotator/package.html]: Mind that the closing tag was already there before, but it was in an illegal position **after** the `</ul>` - all I did is move it before the `<ul>`.

**Comment by mischor** [comment]
I'm ok with these changes.  Can you summarize on the uima-dev list?
I think your guess about extra new lines being caused by the xslt transform code is likely true...  so changing the JAXB stuff wouldn't have an effect.  I wonder if we can change / depend on a particular version of xslt transform code?

File: uimaj-core/src/main/java/org/apache/uima/analysis_engine/annotator/package.html (hunk 3 above)
Comment (mischor) [reply uimaj-core/src/main/java/org/apache/uima/analysis_engine/annotator/package.html]: ok, +1

**Comment by reckart** [comment]
@mischor As I wrote, I tried explicitly depending on Xalan (2.7.2 and 2.7.1) and Xerces (2.9.1) and even Xalan Serializer (2.7.1 and 2.7.2) in the Java9+ profile - to no avail. Note that I assume that Java uses these libraries if they are on the classpath - I didn't explicitly force the use of a specific XML implementation.

**Comment by mischor** [comment]
This was fixed for Java 11 in UIMA-5753.
